packages/base-controller/src/ControllerMessenger.ts

```diff
@@ -53,9 +53,11 @@ export type ExtractEventPayload<
 
 export type GenericEventHandler = (...args: unknown[]) => void;
 
-export type SelectorFunction<Event extends EventConstraint, ReturnValue> = (
-  ...args: ExtractEventPayload<Event>
-) => ReturnValue;
+export type SelectorFunction<
+  Event extends EventConstraint,
+  ReturnValue,
+  EventType = Event['type'],
+> = (...args: ExtractEventPayload<Event, EventType>) => ReturnValue;
```

Comment:
It seems like this can be simplified without breaking any tests? I might just need another explanation on why the `EventType` is needed.

This seems a bit more safe, since I assume we never really want `EventType` to diverge from `Event['type']`.

```suggestion
export type SelectorFunction<
  Event extends EventConstraint,
  ReturnValue,
> = (...args: ExtractEventPayload<Event, Event['type']>) => ReturnValue;
```

Associated changes:

```diff
diff --git a/packages/base-controller/src/ControllerMessenger.ts b/packages/base-controller/src/ControllerMessenger.ts
index ef350cd65..eab2fbf35 100644
--- a/packages/base-controller/src/ControllerMessenger.ts
+++ b/packages/base-controller/src/ControllerMessenger.ts
@@ -322,13 +320,13 @@ export class ControllerMessenger<
   subscribe<EventType extends Event['type'], SelectorReturnValue>(
     eventType: EventType,
     handler: SelectorEventHandler<SelectorReturnValue>,
-    selector: SelectorFunction<Event, SelectorReturnValue, EventType>,
+    selector: SelectorFunction<Event, SelectorReturnValue>,
   ): void;
 
   subscribe<EventType extends Event['type'], SelectorReturnValue>(
     eventType: EventType,
     handler: ExtractEventHandler<Event, EventType>,
-    selector?: SelectorFunction<Event, SelectorReturnValue, EventType>,
+    selector?: SelectorFunction<Event, SelectorReturnValue>,
   ): void {
     let subscribers = this.#events.get(eventType);
     if (!subscribers) {
diff --git a/packages/base-controller/src/RestrictedControllerMessenger.ts b/packages/base-controller/src/RestrictedControllerMessenger.ts
index 49ec88e3f..12a7df5f5 100644
--- a/packages/base-controller/src/RestrictedControllerMessenger.ts
+++ b/packages/base-controller/src/RestrictedControllerMessenger.ts
@@ -275,7 +275,7 @@ export class RestrictedControllerMessenger<
   >(
     eventType: EventType,
     handler: SelectorEventHandler<SelectorReturnValue>,
-    selector: SelectorFunction<Event, SelectorReturnValue, EventType>,
+    selector: SelectorFunction<Event, SelectorReturnValue>,
   ): void;
 
   subscribe<
@@ -286,7 +286,7 @@ export class RestrictedControllerMessenger<
   >(
     event: EventType,
     handler: ExtractEventHandler<Event, EventType>,
-    selector?: SelectorFunction<Event, SelectorReturnValue, EventType>,
+    selector?: SelectorFunction<Event, SelectorReturnValue>,
   ) {
     if (!this.#isAllowedEvent(event)) {
       throw new Error(`Event missing from allow list: ${event}`);


```

Reply:
Hmm. Your changes work, but I don't know why.

The way I read the existing code is this: `subscribe` takes an event type. This event type must be one of the event types that has been used to instantiate the messenger. It would follow, then, that if `subscribe` is given a selector function, its arguments should be the same as the handler for the event. For instance, say my controller is called `CountController` and has a `count` state property. It publishes a `CountController:countChanged` event with a payload of a single argument, the value of `count`. When subscribing to this event and passing just an event handler, then, the handler would receive this payload. However, when passing an event handler and a selector, the selector would receive this payload.

It seems that by passing `Event` and not `EventType` to `SelectorFunction`, however, we are saying that the arguments to the `selector` will be typed using all of the events, not just the one we're interested in. In other words, the type of the arguments to the selector function when subscribing to `CountController:countChanged` would be a union of the payload for both `CountController:countChanged` and `CountController:stateChange`. That's why I thought we needed to pass `EventType`, so we can target the selector function to the event that we have subscribed to.

Is this true or am I reading this wrong?

Reply:
Ah thanks for clearing that up again. I think you're right, and in that case we should ideally make two breaking changes here:

1. Constrain 'EventType' with 'extends Event['type']'
2. Remove the default argument ' = Event['type']'. If the consumer wants that as a fallback value, they should provide it themselves. I don't think we want EventType to be an optional parameter. This would be a pretty disruptive change though.

Reply:
It could work either way, depending on what we tend to pass in for `Event`. But the way you've done it here makes more sense because in this class we typically provide all events as the `Event` parameter.

Reply:
@MajorLift I've made some changes in [f281d49](https://github.com/MetaMask/core/pull/4012/commits/f281d49f4c73a3da4bd6338faed1f367a04c9bd1), does that fit with what you had in mind?